### ISSUE TITLE
Add an option to save scene dependencies paths as relative to the scene

### ISSUE THREAD

**Issue by me2beats** [godotengine/godot-proposals#2327]
### Describe the project you are working on
Plugins

### Describe the problem or limitation you are having in your project
When I copy the scene `res://scene.tscn` with the script `res://script.gd` to another project and add it to a subfolder say `res://test`, I get errors:
![scene loading errors](https://user-images.githubusercontent.com/16458555/108612713-ebd58280-740c-11eb-9383-4ff76d8458ac.png)

I find this annoying. I want to copy scenes from one project to another in any subfolder without a need to fix any dependencies

There is no 100% working way to solve it automatically if it happened.

Also see below which other problems can be solved

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Saving resource paths as relative paths to the scene folder would solve the problem.
- This will definitely make the scenes with resources more autonomous and self-contained
- Scenes would not depend on the path to the scene folder.
- Scenes would not depend on the name of the scene folder.
- This would be convenient for the plugin/addon system.
For example this would expand the possibilities for adding
addons with a complex file structure which does not fit into the generally accepted structure.
- this would be also useful for modding games/apps and other things that require scene files to be as self-contained as possible.
- relative paths for nodes like
```gdscript
onready var node = $Node
```
have already proven to be useful and more "portable" than long scene root based paths, so let's do that for scene paths as well!

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
1. Godot can already read relative paths:

`scene.tscn`:
```gdscript
[ext_resource path="script.gd"]
```
works
but when resaving the scene it becomes absolute:
```gdscript
[ext_resource path="res://script.gd"]
```
So it could be an option to just save these paths as relative (trim up to the scene folder path).

2. node path relative system allows using parents like 
```gdscript
var node = $"../"..
```
I think there's no need in this for file paths (might be discussed tho)

3. Relative resource paths should be optional:
if you want relative paths for scene resources, you just specify that in the editor.

I think it would be fine enough to choose whether the scene uses relative or "absolute" (relative to `res://`) paths, this would mean that all resources in the scene are stored either as relative to scene or relative to `res://`, nothing in between (because it might be harder to implement only some paths to be relative to scene).

So we could have a `Resources Relative Paths` checkbox in the `Scene` main menu (the last item in the image):
![rela](https://user-images.githubusercontent.com/16458555/108615621-da4da400-7427-11eb-98af-b7b3f0483a56.png)

4. A tscn file should store info if it uses relative to scene or relative to `res://` paths

### If this enhancement will not be used often, can it be worked around with a few lines of script?
You could manually fix dependencies, ensure if paths have collisions etc, which is not handy.

### Is there a reason why this should be core and not an add-on in the asset library?
Loading and saving scenes related

**Comment by me2beats** [comment]
Related Issues:
- Add relative path support:
https://github.com/godotengine/godot/issues/22019
- Relative paths in tscn files work but godot substitutes with absolute ones on save:
https://github.com/godotengine/godot/issues/24055

**Comment by LightningAA** [comment]
Won't this be solved by the new add-on/sub-project system in Godot 4.0?

**Comment by me2beats** [comment]
> Won't this be solved by the new add-on/sub-project system in Godot 4.0?

Not very familiar with the new addon system, maybe it could solve these problems for plugins
however I do not think that it would solve the problem in general.

---

Btw it looks like a similar feature to `relative paths` has already been used once in Godot. But then it was excluded.
https://github.com/godotengine/godot/issues/24055#issuecomment-444906754
> Relative path caused some trouble back then, so they were removed.

I think any Info on the reasons for abandoning this feature would be helpful.

**Comment by Xrayez** [comment]
> > Won't this be solved by the new add-on/sub-project system in Godot 4.0?
> 
> Not very familiar with the new addon system, maybe it could solve these problems for plugins
> however I do not think that it would solve the problem in general.

Yeah, the idea is similar but not the same. The sub-project system doesn't necessarily involve plugin/addon system. I've actually got the subproject part working if you look at https://github.com/godotengine/godot-proposals/issues/1205#issuecomment-758661586, but this is only useful if you have nested projects you'd like to edit/distribute in a self-contained way. But paths to resources still remain absolute (starting with `res://`), and it sort of works for my use case.

The subproject and plugin systems can (or rather should) be treated separately on conceptual level. But yeah, the new subproject system could greatly improve the way plugins are currently managed (would also remove the need to have plugins to be detected recursively in `res://addons/` recently implemented in godotengine/godot#43734, since they could be found anywhere within a main project).

---

> Btw it looks like a similar feature to `relative paths` has already been used once in Godot. But then it was excluded.
> [godotengine/godot#24055 (comment)](https://github.com/godotengine/godot/issues/24055#issuecomment-444906754)
> 
> > Relative path caused some trouble back then, so they were removed.
> 
> I think any Info on the reasons for abandoning this feature would be helpful.

Probably because the engine assumes absolute paths to be encoded for resources, that's a reliable way they can be currently identified as unique. But this problem could be solved with another approach described in godotengine/godot#15673.

**Comment by me2beats** [comment]
> Won't this be solved by the new add-on/sub-project system in Godot 4.0?
sub-projects are planned in Godot 5 now.

Well the general problem is to be able to easily copy-paste scenes with all dependencies.
There are several ways to solve it.
Another way would be an ability to make scene packages from any scene.

**Comment by KoBeWi** [comment]
I think the uids solved this problem.

**Comment by Mickeon** [comment]
Yeah, I feel like UIDS do what the proposal needs unique paths for, and more.

**Comment by YeldhamDev** [comment]
Not really for scripts (and I assume any other file that isn't a Godot resource), as they error out if not in the exact path they were defined.

**Comment by KoBeWi** [comment]
There is a proposal for script uids: 4603
Other non-resource files can't be dependencies and you have to manage them yourself, so they don't have this problem.

**Comment by YeldhamDev** [comment]
@KoBeWi
> Other non-resource files can't be dependencies

What do you mean? They are dependencies, and if they're moved from the original path (even if together with the scene file) they will break.

> and you have to manage them yourself, so they don't have this problem.

The point of this proposal is to have those paths be relative, so we don't need to manage those files ourselves. :P

**Comment by KoBeWi** [comment]
Ah I was referring to json/txt etc files, but they are a TextFile resource now, so nevermind actually :/

They are similar to scripts though, i.e. you can just put UID in their first line. But some types need a special syntax for that, e.g. cfg files would become invalid if the UID line isn't a comment. But it's not like UID needs to be the first thing in the line 🤔

**Comment by YeldhamDev** [comment]
The problem is files that can't have any metadata stored in them, such as images and sounds. I assume the solution would be to store the UID in their *.export files.

**Comment by KoBeWi** [comment]
Scripts have UIDs now, so this is resolved I think.

Also
> Ah I was referring to json/txt etc files, but they are a TextFile resource now, so nevermind actually :/

idk if it was true then, but text files are no longer considered Resources and can't be tracked via dependencies, so this is irrelevant to the proposal.
